### PR TITLE
Make set() a serializable data type

### LINKED ISSUE #4932: TypeError: set([...]) is not JSON serializable
If I insert a set instead of a list into some plotting function, like `patches` or `line` I get the following error, and after that bokeh shows it even if I use a list and python interpreter has to be restarted to work again properly:

> TypeError                                 Traceback (most recent call last)
> <ipython-input-2-2f9c7f60120a> in <module>()
>       8 p = figure(title="example",webgl=True)
>       9 p.line([280.0, 290.0, 300.0],[280.0, 290.0, 300.0], alpha=0.2)
> ---> 10 show(p)
> 
> c:\Python27\lib\site-packages\bokeh\io.pyc in show(obj, browser, new)
>     299 
>     300     '''
> --> 301     return _show_with_state(obj, _state, browser, new)
>     302 
>     303 def _show_with_state(obj, state, browser, new):
> 
> c:\Python27\lib\site-packages\bokeh\io.pyc in _show_with_state(obj, state, browser, new)
>     307 
>     308     if state.notebook:
> --> 309         comms_handle = _show_notebook_with_state(obj, state)
>     310 
>     311     elif state.server_enabled:
> 
> c:\Python27\lib\site-packages\bokeh\io.pyc in _show_notebook_with_state(obj, state)
>     328     else:
>     329         comms_target = make_id()
> --> 330         publish_display_data({'text/html': notebook_div(obj, comms_target)})
>     331         handle = _CommsHandle(get_comms(comms_target), state.document, state.document.to_json())
>     332         state.last_comms_handle = handle
> 
> c:\Python27\lib\site-packages\bokeh\embed.pyc in notebook_div(model, notebook_comms_target)
>     247 
>     248     with _ModelInDocument(model):
> --> 249         (docs_json, render_items) = _standalone_docs_json_and_render_items([model])
>     250 
>     251     item = render_items[0]
> 
> c:\Python27\lib\site-packages\bokeh\embed.pyc in _standalone_docs_json_and_render_items(models)
>     564     docs_json = {}
>     565     for k, v in docs_by_id.items():
> --> 566         docs_json[k] = v.to_json()
>     567 
>     568     return (docs_json, render_items)
> 
> c:\Python27\lib\site-packages\bokeh\document.pyc in to_json(self)
>     730         # this is a total hack to go via a string, needed because
>     731         # our BokehJSONEncoder goes straight to a string.
> --> 732         doc_json = self.to_json_string()
>     733         return loads(doc_json)
>     734 
> 
> c:\Python27\lib\site-packages\bokeh\document.pyc in to_json_string(self, indent)
>     723         }
>     724 
> --> 725         return serialize_json(json, indent=indent)
>     726 
>     727     def to_json(self):
> 
> c:\Python27\lib\site-packages\bokeh\core\json_encoder.pyc in serialize_json(obj, encoder, indent, **kwargs)
>      95         indent = 2
>      96 
> ---> 97     return json.dumps(obj, cls=encoder, allow_nan=False, indent=indent, separators=separators, sort_keys=True, **kwargs)
> 
> c:\Python27\lib\json\__init__.pyc in dumps(obj, skipkeys, ensure_ascii, check_circular, allow_nan, cls, indent, separators, encoding, default, sort_keys, **kw)
>     248         check_circular=check_circular, allow_nan=allow_nan, indent=indent,
>     249         separators=separators, encoding=encoding, default=default,
> --> 250         sort_keys=sort_keys, **kw).encode(obj)
>     251 
>     252 
> 
> c:\Python27\lib\json\encoder.pyc in encode(self, o)
>     207         chunks = self.iterencode(o, _one_shot=True)
>     208         if not isinstance(chunks, (list, tuple)):
> --> 209             chunks = list(chunks)
>     210         return ''.join(chunks)
>     211 
> 
> c:\Python27\lib\json\encoder.pyc in _iterencode(o, _current_indent_level)
>     432                 yield chunk
>     433         elif isinstance(o, dict):
> --> 434             for chunk in _iterencode_dict(o, _current_indent_level):
>     435                 yield chunk
>     436         else:
> 
> c:\Python27\lib\json\encoder.pyc in _iterencode_dict(dct, _current_indent_level)
>     406                 else:
>     407                     chunks = _iterencode(value, _current_indent_level)
> --> 408                 for chunk in chunks:
>     409                     yield chunk
>     410         if newline_indent is not None:
> 
> c:\Python27\lib\json\encoder.pyc in _iterencode_dict(dct, _current_indent_level)
>     406                 else:
>     407                     chunks = _iterencode(value, _current_indent_level)
> --> 408                 for chunk in chunks:
>     409                     yield chunk
>     410         if newline_indent is not None:
> 
> c:\Python27\lib\json\encoder.pyc in _iterencode_list(lst, _current_indent_level)
>     330                 else:
>     331                     chunks = _iterencode(value, _current_indent_level)
> --> 332                 for chunk in chunks:
>     333                     yield chunk
>     334         if newline_indent is not None:
> 
> c:\Python27\lib\json\encoder.pyc in _iterencode_dict(dct, _current_indent_level)
>     406                 else:
>     407                     chunks = _iterencode(value, _current_indent_level)
> --> 408                 for chunk in chunks:
>     409                     yield chunk
>     410         if newline_indent is not None:
> 
> c:\Python27\lib\json\encoder.pyc in _iterencode_dict(dct, _current_indent_level)
>     406                 else:
>     407                     chunks = _iterencode(value, _current_indent_level)
> --> 408                 for chunk in chunks:
>     409                     yield chunk
>     410         if newline_indent is not None:
> 
> c:\Python27\lib\json\encoder.pyc in _iterencode_dict(dct, _current_indent_level)
>     406                 else:
>     407                     chunks = _iterencode(value, _current_indent_level)
> --> 408                 for chunk in chunks:
>     409                     yield chunk
>     410         if newline_indent is not None:
> 
> c:\Python27\lib\json\encoder.pyc in _iterencode(o, _current_indent_level)
>     440                     raise ValueError("Circular reference detected")
>     441                 markers[markerid] = o
> --> 442             o = _default(o)
>     443             for chunk in _iterencode(o, _current_indent_level):
>     444                 yield chunk
> 
> c:\Python27\lib\site-packages\bokeh\core\json_encoder.pyc in default(self, obj)
>      79             return obj.to_css()
>      80         else:
> ---> 81             return self.transform_python_types(obj)
>      82 
>      83 def serialize_json(obj, encoder=BokehJSONEncoder, indent=None, **kwargs):
> 
> c:\Python27\lib\site-packages\bokeh\core\json_encoder.pyc in transform_python_types(self, obj)
>      60             return float(obj)
>      61         else:
> ---> 62             return super(BokehJSONEncoder, self).default(obj)
>      63 
>      64     def default(self, obj):
> 
> c:\Python27\lib\json\encoder.pyc in default(self, o)
>     182 
>     183         """
> --> 184         raise TypeError(repr(o) + " is not JSON serializable")
>     185 
>     186     def encode(self, o):
> 
> TypeError: set([280.0, 290.0, 300.0]) is not JSON serializable
> 

```

import bokeh
from bokeh.plotting import figure, output_notebook, show
output_notebook(resources=bokeh.resources.INLINE)

p = figure(title="example",webgl=True)
p.patches({280.0, 290.0, 300.0},{280.0, 290.0, 300.0}, alpha=0.2)
show(p)
```

bokeh version: 0.11.1

### PR DESCRIPTION
**Pull request by mattpap**
fixes #4932

### CODE REVIEW

File: bokeh/util/serialization.py [hunk 1]
```diff
@@ -112,11 +112,16 @@ def traverse_data(datum, is_numpy=is_numpy, use_numpy=True):
         use_numpy: toggle numpy as a dependency for testing purposes
     """
     is_numpy = is_numpy and use_numpy
-    if is_numpy and not any(isinstance(el, (list, tuple)) for el in datum):
+    if is_numpy and not any(isinstance(el, (list, tuple, set)) for el in datum):
+        # NOTE: this is necessary, because later .tolist() is used and it will
+        # keep the original, non-serializable set type. Yes, .tolist() doesn't
```
Comment (birdsarah) [inline bokeh/util/serialization.py]: Well you learn something new every day!

**Comment by bryevdv** [comment]
Sets are unordered. Doesn't make sense at all to serialize them?

**Comment by bryevdv** [comment]
@birdsarah @mattpap what are the use cases for serializing `set`? They are unordered, so they are unsuitable for columns in a `ColumnDataSource`, etc. I could possibly see a separate `Set` property type but I think the places where serializing `set` makes sense are few. In particular, something like:
```
p.circle({1,2,3}, ...)
```
Should just be an error. Am I missing something?

File: bokeh/util/serialization.py (hunk 1 above)
Comment (bryevdv) [inline bokeh/util/serialization.py]: I'm confused. `.tolist` is a method on `ndarray`, not `set`.  Just for my own edification can you provide a minimal test case that demonstrates this claim?

**Comment by birdsarah** [comment]
> @birdsarah @mattpap what are the use cases for serializing `set`?

I've never had a need to, and I can't think of anything off the top of my head.

**Comment by bryevdv** [comment]
I'm inclined to close this, then.

**Comment by bryevdv** [comment]
As no-one has justified using sets in this way, I'm going to close this PR. I will leave the issue open as a task to add much better error reporting than there is currently.

**Comment by mattpap** [comment]
I see there is confusion here about the difference between being able to serialize instances of `set` type and allowing sets to be used as plotable collections. This PR is about the former. I'm not convinced if with need this after all, but just for serializer's completeness it would be beneficial.

As this PR was closed, a new issue, disallowing sets in `ColumnDataSource.data` and possibly appropriate APIs should have been created, or the original issue adapted. There is no need for fancy error reporting anywhere around this issue, just `ColumnDataSource.data`'s type should be changed to `Dict(String, OrderedSeq)`, which appropriate definition of `OrderedSeq` (perhaps something from `collections` module would match).

**Comment by mattpap** [comment]
btw. `ColumnDataSource.data`'s type should be changed anyway, because `Any` is too broad. E.g. we definitively don't allow scalars there, yet it's currently a perfectly acceptable type.

**Comment by bryevdv** [comment]
Agree that sounds like a reasonable approach to better validation and error reporting using the facilities already in place.

**Pull request closed without merging**